AssemblerLib/MeshComponentMap.cpp

```diff
@@ -34,67 +34,86 @@ MeshComponentMap::MeshComponentMap(
-                                                          + mesh.getGlobalNodeID(j) );
-                    const GlobalIndexType signed_global_id = mesh.isGhostNode( mesh.getNode(j)->getID() )
-                                                    ? -global_id : global_id;
-                    _dict.insert(Line(Location(mesh_id, MeshLib::MeshItemType::Node, j),
-                                comp_id, signed_global_id) );
-                }
-
-                // Note: If the cells are really used (e.g. for the mixed FEM),
-                // the following global cell index must be reconsidered
-                // according to the employed cell indexing method.
-                for (std::size_t j=0; j<mesh_subset.getNElements(); j++)
-                    _dict.insert(Line(Location(mesh_id, MeshLib::MeshItemType::Cell, j),
-                                 comp_id, cell_index++));
-                // Include base nodes.
-                global_index_offset += mesh.getNGlobalNodes();
-            }
-
-            _num_global_dof += mesh.getNGlobalNodes();
-       }
+                 GlobalIndexType global_id = 0;
+                 if (order == ComponentOrder::BY_LOCATION)
+                 {
+                    global_id = static_cast<GlobalIndexType>(components.size()
+                                   * mesh.getGlobalNodeID(j) + comp_id);
+                 }
+                 else
+                 {
+                     global_id = static_cast<GlobalIndexType>(global_index_offset
+                                             + mesh.getGlobalNodeID(j) );
+                 }
+                 const bool is_ghost = mesh.isGhostNode( mesh.getNode(j)->getID() );
+                 if (is_ghost)
+                 {
+                     _ghosts_indices.push_back(global_id);
+                     global_id = -global_id;
+                     // If the ghost entry has an index of 0,
+                     // its index is set to the negative value of unknowns.
+                     if (global_id == 0)
+                          global_id = -num_unknowns;
+                 }
+                 else
+                     _num_local_dof++;
+                       
+                 _dict.insert(Line(Location(mesh_id, MeshLib::MeshItemType::Node, j),
+                                comp_id, global_id) );
+             }
+
+             // Note: If the cells are really used (e.g. for the mixed FEM),
+             // the following global cell index must be reconsidered
+             // according to the employed cell indexing method.
+             for (std::size_t j=0; j<mesh_subset.getNElements(); j++)
+                 _dict.insert(Line(Location(mesh_id, MeshLib::MeshItemType::Cell, j),
+                              comp_id, cell_index++));
+
+             if (order == ComponentOrder::BY_COMPONENT)
+             {
+                  // Include base nodes.
+                  global_index_offset += mesh.getNGlobalNodes();
```

Comment:
It looks like `global_index_offset` could be replaced by `comp_id * mesh.getNGlobalNodes()` or by `_num_global_dof`. That would simplify the code a bit.